cmd/dockerd/config.go

```diff
@@ -64,6 +64,7 @@ func installCommonConfigFlags(conf *config.Config, flags *pflag.FlagSet) error {
 	flags.Var(opts.NewListOptsRef(&conf.DNS, opts.ValidateIPAddress), "dns", "DNS server to use")
 	flags.Var(opts.NewNamedListOptsRef("dns-opts", &conf.DNSOptions, nil), "dns-opt", "DNS options to use")
 	flags.Var(opts.NewListOptsRef(&conf.DNSSearch, opts.ValidateDNSSearch), "dns-search", "DNS search domains to use")
+	flags.BoolVar(&conf.DNSResolveDockerHost, "dns-resolve-docker-host", true, "dockerd will resolve the host.docker.internal DNS name if set to true")
```

Comment:
Not a big fan of making this a boolean. Perhaps we could make this flag (that can be specified multiple times) to add custom entries to the resolver (either the embedded DNS or (see my other comment) entries in `/etc/hosts`, depending on the situation).

For this particular case, we may need a "special" value for the gateway address (which I think is dynamically set), so that users can (haven't thought of the name of the flag yet);

```bash
--custom-dns-entry host.docker.internal=gateway \
--custom-dns-entry gateway.docker.internal= gateway
```

With that (perhaps) Docker Desktop could also use that option?

Reply:
This would also make it compatible using Compose's `extra_hosts:` field.

Reply:
that option is per container/service, not at the daemon level

Reply:
@thaJeztah for this case, the community prefers some `automatic`

Reply:
also if we use embedded DNS it won't support this DNS name for containers attached to the default `bridge`/`docker0` network

Reply:
> @thaJeztah for this case, the community prefers some automagic

While I understand that some magic could be nice to have, I think the reason we originally didn't add it on Linux was specifically because hard-coding could cause issues in some situations;

For reference, the `host.docker.internal` was added on Docker Desktop to get parity with Linux (see https://github.com/moby/moby/issues/22753: "No docker0 on docker for mac?"). 

Related discussions use-cases / issues were

- https://github.com/moby/moby/issues/1143 Document how to connect to Docker host from container
- https://github.com/moby/moby/issues/23177 Add dockerhost as an entry in /etc/hosts in all containers
- (slightly related) https://github.com/moby/moby/issues/15086 "Document how to get real remote client ip for service running in container"


I _think_ that;

- For the first issue; on Docker Desktop, docker has (almost) full control over how networking is set-up, so implementing `host.docker.internal` could be done because of that, whereas on Linux, networking configuration can widely differ, so hard-coding values may cause issues in some setups
- For the other issues, expectations "how to connect to the host" widely varied;
    - some people were looking at connecting to a (non-containerised) service running on the same host
    - some people were looking at "how to connect to a service running on the same host" (either containerised or non-containerised) on a "interface", and how to get that IP-address?
    - "public interface" / IP-address is ambiguous in setups where the docker host is behind a proxy (and the docker host itself does not know the public address

To get "magic", docker could still decide to set these options in the default configuration (systemd unit?), but hard-coding could potentially (see above) limit the usefulness if users cannot customise these settings, and if the hard-coded values don't work.

Reply:
@thaJeztah the hard coding can always be removed by setting `dns-resolve-docker-host=false` if someone wants to be change this

Reply:
But for the situations mentioned above, there would be no way to change `host.docker.internal=<hard-coded-ip>` to `host.docker.internal=<correct IP>` to be included in all containers.

Reply:
`dns-resolve-docker-host=false` (In Daemon config) && `--add-host host.docker.internal=<some-IP>` would address your concern

There is no ideal answer (I have raised the same questions), but apparently many developers want parity across OSs with the same Compose file and this PR attempts to do that

Reply:
it would have to be set on every container, every service, so not the same as this feature implements